### PR TITLE
VFB-3 Create modal to read JSON input

### PR DESCRIPTION
**Pull request by ChloeYiuSoftwire**
We have created 3 files as follows:

1. DataViewer: contains the modal itself
2. DataViewerWithButton: contains the example for using the DataViewer component. This is also used in the cypress test so please do not remove this.
3. Cypress test file

- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've linted via `npm run lint`
    - can run `npm run lint_fix` to try and fix as any mistakes automatically as possible!
- [x] Make sure you've tested via `npm run test`

### CODE REVIEW

File: src/components/DataViewer/DataViewer.tsx
Comment (daniel-conlan-softwire) [inline src/components/DataViewer/DataViewer.tsx:95]: Ask Tom, but you might want to replace this using styled-components by creating a custom `<StyledModal>`  component within this file. This might not be needed, and may not be worth the effort, but it's something to ask.

File: src/components/DataViewer/DataViewerWithButton.tsx
Comment (daniel-conlan-softwire) [inline src/components/DataViewer/DataViewerWithButton.tsx:1]: If this is only used for testing, then I would just define the component within the test file itself.

If it's not going to be used within the wider application, then I don't think it should be included as a component.

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [reply src/components/DataViewer/DataViewer.tsx:95]: yep, wherever possible (and it is possible here!) I don't want `style=` set, and I would like `styled-components` to be used.
Potentially ask @SamYap-Softwire / @Var1377 for help, they both did some styled component stuff with modals!

File: src/components/DataViewer/DataViewerWithButton.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewerWithButton.tsx:1]: Hmmmm it is a sample file to show how is data viewer implemented though because it has to pass in some useState stuff from the file using it but maybe I will make it clearer that it is a sample file?

File: src/components/DataViewer/DataViewer.tsx
Comment (Var1377) [inline src/components/DataViewer/DataViewer.tsx:16]: Try and use max or min on the inset property here to try and make it responsive, hopefully that works

File: src/components/DataViewer/DataViewer.tsx
Comment (Var1377) [inline src/components/DataViewer/DataViewer.tsx:41]: is there a cleaner way to do this? the modal can be smaller if there's less data to show

File: src/components/DataViewer/DataViewer.tsx
Comment (Var1377) [inline src/components/DataViewer/DataViewer.tsx:56]: it should work without display: flex and flex-direction: column as you aren't using any other flexbox properties

File: src/components/DataViewer/DataViewer.tsx
Comment (Var1377) [inline src/components/DataViewer/DataViewer.tsx:68]: Not a big deal but now that icons are merged you could use a font-awesome icon

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx:8]: We have added comment as suggested

File: src/components/DataViewer/DataViewer.cy.tsx
Comment (DanWilkins2107) [inline src/components/DataViewer/DataViewer.cy.tsx:30]: Could add a separate test for modal shows correct values?

File: src/components/DataViewer/DataViewer.tsx
Comment (DanWilkins2107) [inline src/components/DataViewer/DataViewer.tsx:101]: Not a big deal, and idk what other people think for this, so maybe get a second opinion, but we've been calling our styled version of things Styled_____ and it would be nice for that to be consistent throughout, obvs if everyone else does your way we should change our bits tho :)

**Review by DanWilkins2107 — COMMENTED** [review]
Good apart from Small Changes

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx:57]: Agreed

**Review by daniel-conlan-softwire — APPROVED** [review]

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx:17]: Seems like max and min isn't necessary to make it responsive (?) but we removed those px-dependent number and stick to percentage, and it is responsive.

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx]: The modal now is responsive to the page size and the amount of data to show. We cleaned it with display flex.

File: cypress/e2e/accessibility.cy.ts
Comment (TomCusack-Softwire) [inline cypress/e2e/accessibility.cy.ts:1]: looks like the rebase didn't really work out - these changes shouldn't be on your branch!
they should be on main (which would be 'beneath' your branch), so the changes would be present, but not as an addition or as separate commits...
May need an interactive rebase (to squash some commits or delete some commits) to fix this up!

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [reply src/components/DataViewer/DataViewer.tsx:8]: I don't think this is necessarily needed - I wouldn't expect a nested JSON to be rendered in a single-level table at all, and the type errors 'are' the sign that it won't work! I think specifying that data is a key-value (where value is string/num/null) should be sufficient.

(Especially since we are not exporting this component or making others use it, there won't be 'downstream' users using it for weird edge cases - the component is made in a way which is intuitive and easy-to-use, and it's on the caller to format their data to fit what the table is expecting!)

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [reply src/components/DataViewer/DataViewer.tsx:67]: if you do then please make a new component for it! (this is a purely optional change, I think the X is fine!)

Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx:69]: We will deal with this later.

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx:8]: Ok. We have removed the comment then

File: src/components/DataViewer/DataViewer.cy.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.cy.tsx:30]: We have added the test as suggested in the newer version

File: src/components/DataViewer/DataViewer.tsx
Comment (ChloeYiuSoftwire) [reply src/components/DataViewer/DataViewer.tsx]: We have fixed it as suggested

File: cypress/e2e/accessibility.cy.ts
Comment (ChloeYiuSoftwire) [reply cypress/e2e/accessibility.cy.ts]: resolved as suggested

File: package.json
Comment (TomCusack-Softwire) [inline package.json:76]: can this be all in dependencies (rather than dev)? we're not splitting dependencies for the time being :)

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:99]: what's `ariaHideApp` do?

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:100]: Where does this get shown? Can it be in less 'tech' language - e.g. 'data viewer'?

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:69]: can `<{ closeModal: () => void }>` be extracted into a `ClearButtonProps` please? :)

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]

File: package.json
Comment (paopaosoftwire) [reply package.json:76]: I have moved it

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:99]: "It is important for users of screenreaders that other page content be hidden (via the aria-hidden attribute) while the modal is open." 
https://reactcommunity.org/react-modal/accessibility/

I think it doesn't hurt to include the line in the code.

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:100]: contentLabel is used for accessibility reasons. Yes, I agree that the text should be changed and it has now been changed.

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/SampleDataViewerWithButton.tsx:25]: Removed!

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:69]: Done!

File: src/app/clients/page.tsx
Comment (TomCusack-Softwire) [inline src/app/clients/page.tsx:21]: Can you remove this line? It shouldn't be on the final page (yet) - we're just making the component!

File: src/components/DataViewer/DataViewer.cy.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.cy.tsx:46]: should cy.contains an empty string for dietary requirements be a test? or even that dietary requirements is shown on the page too?

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:68]: can you extract this to an interface?

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:78]: probably also better to have DataViewerProps to be defined right above this, so you don't have to scroll too far to see what props should have in it!

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [reply src/components/DataViewer/DataViewer.tsx:89]: I think it's great, I just didn't know what it did!
On that page, it looks like you should give it the root element rather than 'false':

> To allow react-modal to do this, you should call Modal.setAppElement with a query selector identifying the root of your app

```
Modal.setAppElement('#root');
```

> If you are already applying the aria-hidden attribute to your app content through other means, you can pass the ariaHideApp={false} prop to your modal to avoid getting a warning that your app element is not specified.

so I think with this you are actually turning off the warning?

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:4]: remove the comments please!

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:33]: might be worth using `Button` (the styled one) from the shared components folder

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]

File: src/components/DataViewer/DataViewer.cy.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.cy.tsx:46]: An empty string cannot be passed into cy.contains so I think a better test would be just to test if cy.contains("DIETARY REQUIREMENTS") which should be the header for that section.

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:68]: Done!

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:78]: Done!

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (SamYap-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:7]: the type of data is also used in the dataviewer file. Maybe you can export the interface?

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (SamYap-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:38]: I'm not sure what this comment is referring to

**Review by SamYap-Softwire — CHANGES_REQUESTED** [review]

File: src/components/DataViewer/DataViewer.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.tsx:89]: You are completely correct. We have now added the setAppElement and set ariaHideApp to true. Thanks.

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/SampleDataViewerWithButton.tsx:33]: The current button takes a href which is not what we want as we need to use the onClick. I think we could change it once theres a better button implemented in a separate ticket.

File: src/components/DataViewer/DataViewer.cy.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/DataViewer.cy.tsx:2]: DONE

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/SampleDataViewerWithButton.tsx:7]: DONE

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (paopaosoftwire) [reply src/components/DataViewer/SampleDataViewerWithButton.tsx:38]: It wasn't supposed to be commented out. Thanks!

File: src/components/DataViewer/DataViewer.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/DataViewer.tsx:86]: can this have type `closeModel = (): void => {}`? typing the return (and parameters) individually rather than typing the entire function

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:23]: same here

File: src/components/DataViewer/SampleDataViewerWithButton.tsx
Comment (TomCusack-Softwire) [inline src/components/DataViewer/SampleDataViewerWithButton.tsx:5]: @/ import please :)

**Review by TomCusack-Softwire — APPROVED** [review]
Very small changes, but otherwise I'm happy.
Please do wait for @SamYap-Softwire to also approve before merging! 🙂

File: package.json
Comment (SamYap-Softwire) [inline package.json:29]: image

We want the bottom install instead of the top for styled components

**Review by SamYap-Softwire — APPROVED** [review]
I think it's good.

File: package.json
Comment (paopaosoftwire) [reply package.json:29]: Looking at other branches which has MUI installed, such as navbar, @mui/styled-engine-sc seems to not be used at all. And when installing it via npm there are dependency errors. uninstalling everything and keeping only @mui/material and styled-components seems to work fine so it will be kept like that.

**Pull request merged**
